### ISSUE TITLE
[CPU] Replace simplified parallel attention implementation with proper attention mechanism

### ISSUE THREAD

**Issue by EffortlessSteven** [EffortlessMetrics/BitNet-rs#415]
# [CPU] Replace simplified parallel attention implementation with proper attention mechanism

## Problem Description

The `cpu_optimizations::parallel_attention` function in `crates/bitnet-inference/src/cpu.rs` contains a simplified implementation that lacks proper softmax normalization and complete attention mechanism semantics. The current implementation applies raw exponentials without normalization across sequence positions, which produces incorrect attention weights and compromises inference quality.

## Environment

- **File**: `crates/bitnet-inference/src/cpu.rs`
- **Function**: `cpu_optimizations::parallel_attention` (lines 427-471)
- **Architecture**: CPU backend with Rayon parallelism
- **Build**: `cargo build --no-default-features --features cpu`

## Current Implementation Analysis

```rust
// Current simplified implementation (lines 458-463)
// Apply softmax and compute weighted sum (simplified)
let weight = score.exp(); // Simplified softmax - MISSING NORMALIZATION
for d in 0..head_dim {
    head_output[i * head_dim + d] +=
        weight * value[v_offset + j * head_dim + d];
}
```

**Root Cause Issues:**
1. **Missing Softmax Normalization**: Uses raw `score.exp()` without dividing by sum of exponentials
2. **No Numerical Stability**: Lacks max subtraction for numerical stability in softmax
3. **Incomplete Attention Semantics**: Does not ensure attention weights sum to 1.0 per query position
4. **Accumulation Without Reset**: `head_output` accumulates without proper initialization

## Impact Assessment

**Severity**: High - Affects inference accuracy and model behavior
**Affected Components**:
- CPU backend inference quality
- Parallel attention computation accuracy
- Cross-validation with reference implementations
- Production inference reliability

**Performance Impact**: Current implementation may appear faster but produces incorrect results

## Reproduction Steps

1. Build with CPU features: `cargo build --no-default-features --features cpu`
2. Run attention computation with multiple sequence positions
3. Observe that attention weights do not sum to 1.0 per query position
4. Compare with proper attention implementation in `crates/bitnet-inference/src/layers/attention.rs`

**Expected Result**: Attention weights should sum to 1.0 per query position
**Actual Result**: Raw exponentials used without normalization

## Proposed Solution

Replace the simplified implementation with a proper multi-head attention mechanism:

### Primary Implementation

```rust
pub fn parallel_attention(
    query: &[f32],
    key: &[f32],
    value: &[f32],
    output: &mut [f32],
    seq_len: usize,
    head_dim: usize,
    num_heads: usize,
) -> Result<()> {
    // Initialize output to zeros
    output.fill(0.0);

    // Parallel processing by attention heads
    output
        .par_chunks_mut(seq_len * head_dim)
        .enumerate()
        .try_for_each(|(head_idx, head_output)| -> Result<()> {
            if head_idx >= num_heads {
                return Ok(());
            }

            let q_offset = head_idx * seq_len * head_dim;
            let k_offset = head_idx * seq_len * head_dim;
            let v_offset = head_idx * seq_len * head_dim;

            // Compute attention for each query position
            for i in 0..seq_len {
                // Compute attention scores for query position i
                let mut scores = vec![0.0f32; seq_len];
                let mut max_score = f32::NEG_INFINITY;

                // Calculate raw attention scores
                for j in 0..seq_len {
                    let mut score = 0.0f32;
                    for d in 0..head_dim {
                        score += query[q_offset + i * head_dim + d]
                               * key[k_offset + j * head_dim + d];
                    }

                    // Scale by sqrt(head_dim) for numerical stability
                    score /= (head_dim as f32).sqrt();
                    scores[j] = score;
                    max_score = max_score.max(score);
                }

                // Apply numerically stable softmax
                let mut exp_sum = 0.0f32;
                for score in &mut scores {
                    *score = (*score - max_score).exp();
                    exp_sum += *score;
                }

                // Normalize to get attention weights
                for score in &mut scores {
                    *score /= exp_sum;
                }

                // Compute weighted sum of values
                for d in 0..head_dim {
                    let mut weighted_value = 0.0f32;
                    for j in 0..seq_len {
                        weighted_value += scores[j] * value[v_offset + j * head_dim + d];
                    }
                    head_output[i * head_dim + d] = weighted_value;
                }
            }

            Ok(())
        })?;

    Ok(())
}
```

### Alternative Approach: Memory-Optimized Version

For large sequences, implement a memory-efficient version that processes chunks of the attention matrix:

```rust
pub fn parallel_attention_chunked(
    query: &[f32],
    key: &[f32],
    value: &[f32],
    output: &mut [f32],
    seq_len: usize,
    head_dim: usize,
    num_heads: usize,
    chunk_size: usize,
) -> Result<()> {
    // Process attention in chunks to reduce memory usage
    // Implementation details...
}
```

### Phase 1: Core Implementation
- [ ] Replace simplified softmax with proper numerically stable implementation
- [ ] Add attention weight normalization ensuring sum equals 1.0
- [ ] Implement proper output initialization
- [ ] Add scaling by sqrt(head_dim)

### Phase 2: Optimization
- [ ] Add SIMD optimizations for score computation when available
- [ ] Implement memory-efficient chunked processing for large sequences
- [ ] Add attention matrix caching for repeated computations

### Phase 3: Testing & Validation
- [ ] Create unit tests verifying attention weight normalization
- [ ] Add cross-validation tests against reference attention implementation
- [ ] Benchmark performance impact of proper vs simplified implementation
- [ ] Add integration tests with full model inference

### Phase 4: Documentation
- [ ] Update function documentation with attention mechanism details
- [ ] Add examples of proper usage in CPU backend
- [ ] Document performance characteristics and memory usage

### Unit Tests
```rust
#[test]
fn test_attention_weights_normalization() {
    // Test that attention weights sum to 1.0 per query position
}

#[test]
fn test_attention_numerical_stability() {
    // Test with large score values to verify numerical stability
}

#[test]
fn test_attention_cross_validation() {
    // Compare with reference attention implementation
}
```

### Integration Tests
- Cross-validation with `bitnet-inference/src/layers/attention.rs`
- Performance regression testing
- Memory usage validation

## Related Issues

This issue is part of a broader effort to eliminate simulation code:
- **Multi-Head Attention Mask**: BitNetAttention simplified mask application (#TBD)
- **Backend Implementations**: Comprehensive backend simulation cleanup (#TBD)
- **Validation Testing**: Cross-validation framework improvements (#TBD)

## Risk Assessment

**Implementation Risks**:
- Performance regression due to proper normalization overhead
- Memory usage increase for intermediate score storage
- Potential numerical precision differences in edge cases

**Mitigation Strategies**:
- Benchmark before/after performance
- Implement memory-efficient chunked processing option
- Add extensive numerical validation tests
- Gradual rollout with feature flags

## Acceptance Criteria

- [ ] Attention weights sum to 1.0 ± 1e-6 per query position
- [ ] Numerically stable softmax implementation with max subtraction
- [ ] Performance within 20% of current implementation
- [ ] Cross-validation tests pass against reference implementation
- [ ] No memory leaks or excessive allocations
- [ ] Proper error handling for edge cases (seq_len=0, invalid dimensions)
- [ ] Documentation updated with attention mechanism details
- [ ] Unit tests achieve >95% code coverage

## Notes

This implementation should align with the attention mechanism in `crates/bitnet-inference/src/layers/attention.rs` while maintaining the parallel processing benefits of the CPU backend. The fix directly impacts the quality of neural network inference and is critical for production readiness.
